### ISSUE TITLE
NER fails when using built-in entity

### ISSUE THREAD

**Issue by obaid** [axa-group/nlp.js#210]
**Describe the bug**
When an utterance just has a built-in entity as a training phrase, the application crashes.

**To Reproduce**

```
manager.addNamedEntityText("27e3ee80-99ec-11e9-9113-bfbeddbabe4c", "asdf", ["en"], ["asdf"] );
manager.addDocument("en", "%27e3ee80-99ec-11e9-9113-bfbeddbabe4c%", "bb969130-99ec-11e9-8271-53f83197e616");
manager.addDocument("en", "sentenence %27e3ee80-99ec-11e9-9113-bfbeddbabe4c% hdi", "bb969130-99ec-11e9-8271-53f83197e616");
manager.addDocument("en", "%email%", "c13aae30-99ec-11e9-9912-a5b098103b51");


manager.train().then(()=>{
  manager.process('en', 'asdf@hodi.com', {}).then(result => console.log(result));
});
```

**Expected behavior**
The result should have intent `c13aae30-99ec-11e9-9912-a5b098103b51` as a result. But it just crashes with this message:

```
(node:80601) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'shift' of undefined
    at edges.map.item (/Users/o/Desktop/nlp/node_modules/node-nlp/lib/ner/ner-manager.js:520:48)
    at Array.map (<anonymous>)
    at NerManager.findEntities (/Users/o/Desktop/nlp/node_modules/node-nlp/lib/ner/ner-manager.js:518:21)
    at NlpManager.process (/Users/o/Desktop/nlp/node_modules/node-nlp/lib/nlp/nlp-manager.js:386:45)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:757:11)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
(node:80601) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:80601) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

If I take this line out it works:

```
manager.addNamedEntityText("27e3ee80-99ec-11e9-9113-bfbeddbabe4c", "asdf", ["en"], ["asdf"] );
```

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: OSX
 - Browser N/A
 - Version N/A
- Node 10.15

**Additional context**

**Comment by jesus-seijas-sp** [comment]
Hello, I see the error, but I have a question: What you expect that it will return?

I will explain, you have "asdf@hodi.com" but "asdf" is also the entity "27e3ee80-99ec-11e9-9113-bfbeddbabe4c", so as a result it will return two entities overlaped: the entity "email" and the entity "27e3ee80-99ec-11e9-9113-bfbeddbabe4c", so when reducing the edges only one of the entities will win, in this case will win "27e3ee80-99ec-11e9-9113-bfbeddbabe4c" because is the one manually defined by you, so has more priority over builtin.

What you expect as result?

**Comment by obaid** [comment]
Ideally since the email is a perfect match I would expect email to win.

**Comment by gabrielvega** [comment]
Hi! I have a similar issue.

 Let me share my code:

```
manager.addNamedEntityText("drink", "soda", ["en"], ["soda"] );
manager.addDocument("en", "may I have a %drink%", "buy_drink");
manager.addDocument("en", "I want a %drink%", "buy_drink");

manager.train().then(()=>{
  manager.process('en', 'may i have a soda', {}).then(result => console.log(result));
});
```

Error message:

> (node:17248) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'shift' of undefined
>     at edges.map.item (C:\Users\gabri\Projects\nlp-node\node_modules\node-nlp\lib\ner\ner-manager.js:520:48)

It turns out, that the word may (month) is the one that it's creating the issue.

Any thoughts?

Thanks in advance.

**Comment by jesus-seijas-sp** [comment]
Error fixed in version 3.2.1, now does not throw the exception and return entities, but edges reduced right now with preference for manually introduced named entities over builtin.
Can you confirm?

**Comment by gabrielvega** [comment]
> Error fixed in version 3.2.1, now does not throw the exception and return entities, but edges reduced right now with preference for manually introduced named entities over builtin.
> Can you confirm?

Confirmed!. Both errors that you mentioned  were fixed.
This test may not have too much sense but it shows the preference for manually over builtin entities...

```
const { NlpManager, ConversationContext } = require('node-nlp');

var context = new ConversationContext();
const manager = new NlpManager({ languages: 'en' });

manager.addNamedEntityText('drink', 'coke', ['en'], ['coke', 'soda']);
manager.addNamedEntityText('my_entity', 'may', ['en'], ['may']);
manager.addDocument('en', '%my_entity% I have a', 'buy_drink');

(async () => {
    await manager.train();
    var processed = await manager.process('en', 'may i have a soda', context);
    console.log(processed);
})();
```


Output

```
{
  "utterance": "may i have a soda",
  "locale": "en",
  "languageGuessed": false,
  "localeIso2": "en",
  "language": "English",
  "domain": "order.drink",
  "classifications": [
    {
      "label": "buy_drink",
      "value": 1
    }
  ],
  "intent": "buy_drink",
  "score": 1,
  "entities": [
    {
      "start": 0,
      "end": 2,
      "len": 3,
      "levenshtein": 0,
      "accuracy": 1,
      "option": "may",
      "sourceText": "may",
      "entity": "my_entity",
      "utteranceText": "may"
    },
    {
      "start": 13,
      "end": 16,
      "len": 4,
      "levenshtein": 0,
      "accuracy": 1,
      "option": "coke",
      "sourceText": "soda",
      "entity": "drink",
      "utteranceText": "soda"
    }
  ],
  "sentiment": {
    "score": 0,
    "comparative": 0,
    "vote": "neutral",
    "numWords": 5,
    "numHits": 0,
    "type": "senticon",
    "language": "en"
  },
  "actions": []
}
```

**Comment by jesus-seijas-sp** [comment]
Closing due to confirmation of solved.
